File: packages/capabilities-runtime-ts/src/context.ts
import type { CancellationToken } from './cancellation.js'
import type { Clock } from './clock.js'
import type { ConfigurationReader } from './configuration.js'
import type { SecretResolver } from './secrets.js'
import type { Logger, Tracer } from './telemetry.js'

/** Authenticated caller identity, if any. Identity-provider integration is a host concern. */
export interface Principal {
  readonly id?: string
  readonly claims?: Readonly<Record<string, unknown>>
}

/**
 * Authorization hook (§15.2). Protected operations default to deny until a
 * hook is registered; the runtime defines the protocol only, never policy.
 */
export interface AuthorizationHook {
  isAuthorized(operationCode: string, principal: Principal | undefined): boolean | Promise<boolean>
}

/**
 * Per-dispatch execution context (§10.1). Framework-neutral: every field is
 * a protocol/hook, never a concrete Node/browser dependency.
 */
export interface Context {
  readonly correlationId: string
  readonly cancellation: CancellationToken
  /** Epoch-millisecond absolute deadline; absent means no deadline. */
  readonly deadline?: number
  readonly principal?: Principal
  readonly authorization?: AuthorizationHook
  readonly configuration: ConfigurationReader
  readonly secretResolver: SecretResolver
  readonly logger: Logger
  readonly tracer?: Tracer
  /** Injectable clock for deterministic deadline enforcement; defaults to the system clock. */
  readonly clock?: Clock
}

File: runtimes/python/src/engineering_ui_capabilities_runtime/core/context.py
"""`Context` — everything an `Operation.execute` receives besides its input
(§10.1):

    Context
      correlationId
      cancellation
      deadline
      principal/auth context hook
      configuration reader
      secret resolver reference
      logger/tracer
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Optional

from .cancellation import CancellationToken
from .clock import Clock, SystemClock
from .config import ConfigReader, MappingConfigReader
from .deadline import Deadline
from .secrets import SecretResolver
from .telemetry_protocols import Logger, NullLogger, NullTracer, Tracer

AuthHook = Callable[[str], bool]


@dataclass
class Context:
    correlation_id: str
    cancellation: CancellationToken = field(default_factory=CancellationToken)
    deadline: Optional[Deadline] = None
    config: ConfigReader = field(default_factory=MappingConfigReader)
    secrets: Optional[SecretResolver] = None
    logger: Logger = field(default_factory=NullLogger)
    tracer: Tracer = field(default_factory=NullTracer)
    clock: Clock = field(default_factory=SystemClock)
    principal: Optional[Any] = None
    authorize: Optional[AuthHook] = None

    def is_authorized(self, permission: str) -> bool:
        """Protected operations default to deny until a hook is registered
        (§15.2). The runtime provides only this hook shape; identity
        providers and policy engines are host/project concerns.
        """
        if self.authorize is None:
            return False
        return bool(self.authorize(permission))
